src/ucp/proto/proto_multi.c

```diff
@@ -33,103 +33,48 @@ ucp_proto_multi_compare_tl_perf(const void *elem1, const void *elem2, void *arg)
-                                  const char *perf_name,
-                                  ucp_proto_perf_t **perf_p,
-                                  ucp_proto_multi_priv_t *mpriv)
+static ucp_lane_map_t
+ucp_proto_multi_filter_slow_lanes(ucp_lane_index_t *lanes,
+                                  ucp_proto_common_tl_perf_t *lanes_perf,
+                                  ucp_lane_index_t num_lanes,
+                                  ucp_lane_index_t max_lanes,
+                                  double max_bandwidth, double max_bw_ratio)
 {
-    ucp_context_h context         = params->super.super.worker->context;
-    const double max_bw_ratio     = context->config.ext.multi_lane_max_ratio;
-    ucp_proto_perf_node_t *lanes_perf_nodes[UCP_PROTO_MAX_LANES];
-    ucp_proto_common_tl_perf_t lanes_perf[UCP_PROTO_MAX_LANES];
-    ucp_proto_common_tl_perf_t *lane_perf, perf;
-    ucp_lane_index_t lanes[UCP_PROTO_MAX_LANES];
-    double max_bandwidth, max_frag_ratio, min_bandwidth;
-    ucp_lane_index_t i, lane, num_lanes, first_lane;
-    ucp_proto_multi_lane_priv_t *lpriv;
-    ucp_proto_perf_node_t *perf_node;
-    size_t max_frag, min_length, min_end_offset;
-    ucp_lane_map_t lane_map;
-    ucp_md_map_t reg_md_map;
-    uint32_t weight_sum;
-    ucs_status_t status;
-
-    ucs_assert(params->max_lanes <= UCP_PROTO_MAX_LANES);
-
-    if ((ucp_proto_select_op_flags(params->super.super.select_param) &
-         UCP_PROTO_SELECT_OP_FLAG_RESUME) &&
-        !(params->super.flags & UCP_PROTO_COMMON_INIT_FLAG_RESUME)) {
-        return UCS_ERR_UNSUPPORTED;
-    }
-
-    if (!ucp_proto_common_init_check_err_handling(&params->super) ||
-        (params->max_lanes == 0)) {
-        return UCS_ERR_UNSUPPORTED;
-    }
-
-    /* Find first lane */
-    num_lanes = ucp_proto_common_find_lanes_with_min_frag(
-            &params->super, params->first.lane_type, params->first.tl_cap_flags,
-            1, 0, lanes);
-    if (num_lanes == 0) {
-        ucs_trace("no lanes for %s",
-                  ucp_proto_id_field(params->super.super.proto_id, name));
-        return UCS_ERR_NO_ELEM;
-    }
-
-    /* Find rest of the lanes */
-    num_lanes += ucp_proto_common_find_lanes_with_min_frag(
-            &params->super, params->middle.lane_type,
-            params->middle.tl_cap_flags, UCP_PROTO_MAX_LANES - 1,
-            UCS_BIT(lanes[0]), lanes + 1);
+    ucp_lane_map_t lane_map = 0;
+    ucp_proto_common_tl_perf_t *lane_perf;
+    ucp_lane_index_t i, lane;
 
-    /* Get bandwidth of all lanes and max_bandwidth */
-    max_bandwidth = 0;
     for (i = 0; i < num_lanes; ++i) {
         lane      = lanes[i];
         lane_perf = &lanes_perf[lane];
 
-        status = ucp_proto_common_get_lane_perf(&params->super, lane, lane_perf,
-                                                &lanes_perf_nodes[lane]);
-        if (status != UCS_OK) {
-            return status;
+        if (ucs_popcount(lane_map) == max_lanes) {
```

Comment:
minor: maybe?
```C
for (i = 0; (i < num_lanes) && (ucs_popcount(lane_map) < max_lanes); ++i) {
```

Reply:
I intentionally didn't made this by this way since I think the current one is more readable